### PR TITLE
Publish PowerShell `vPack` for stable and better builds

### PR DESCRIPTION
**Pull request by TravisEz13**
# PR Summary

Publish PowerShell vPack for stable and better builds

## PR Context

This is related to windows engineering

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: tools/releaseBuild/azureDevOps/vpackRelease.yml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      & $azcopy /Source:https://$(StorageAccount).blob.core.windows.net/$(AzureVersion) /Dest:$(System.ArtifactsDirectory) /S /SourceKey:$(StorageAccountKey)
+    displayName: 'Download Azure Artifacts'
+
+  - pwsh: 'Get-ChildItem $(System.ArtifactsDirectory)\* -recurse | Select-Object -ExpandProperty Name'
+    displayName: 'Capture Artifact Listing'
+
+  - pwsh: |
+      if('$(ReleaseTagVar)' -like '*-*')
+      {
+          throw "Don't release a vpack for a preview"
```
Comment (SteveL-MSFT) [inline tools/releaseBuild/azureDevOps/vpackRelease.yml]: The rc releases are expected to be "go live" so should be make an exception for those?

File: tools/releaseBuild/azureDevOps/vpackRelease.yml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  - task: PkgESVPack@10
+    displayName: 'Package ES - VPack '
+    env:
+      SYSTEM_ACCESSTOKEN: $(System.AccessToken)
+    inputs:
+      sourceDirectory: '$(vpackDir)'
+      description: PowerShell $(version)
+      pushPkgName: 'PowerShell.$(vpackBranch).PowerShell'
+      configurations: Release
+      platforms: x64
```
Comment (SteveL-MSFT) [inline tools/releaseBuild/azureDevOps/vpackRelease.yml]: Seems like we need to support all the architectures

**Comment by TravisEz13** [comment]
Code is ready for review now.

File: tools/releaseBuild/azureDevOps/vpackRelease.yml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    include:
+    - master
+    - release*
+
+variables:
+  - name: DOTNET_CLI_TELEMETRY_OPTOUT
+    value: 1
+  - name: POWERSHELL_TELEMETRY_OPTOUT
+    value: 1
+  - group: Azure Blob variable group
```
Comment (TravisEz13) [inline tools/releaseBuild/azureDevOps/vpackRelease.yml]: @adityapatwardhan Note that this YAML tells what variable groups it needs

File: tools/releaseBuild/azureDevOps/templates/vpackReleaseJob.yml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                $message += $messageInstance
+                Write-Warning $messageInstance
+          }
+      }
+
+      if($message.count -gt 0){throw ($message | out-string)}
+    displayName: 'Validate Zip and MSI Package Names'
+
+  - pwsh: |
+      $message = @()
```
Comment (adityapatwardhan) [inline tools/releaseBuild/azureDevOps/templates/vpackReleaseJob.yml]: seems $message is not used

File: tools/releaseBuild/azureDevOps/vpackRelease.yml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+resources:
+- repo: self
+  clean: true
+
+jobs:
+- job: rename
+  displayName: Name the build
+  condition: succeeded()
+  pool:
+    vmImage: 'windows-2019'
```
Comment (adityapatwardhan) [inline tools/releaseBuild/azureDevOps/vpackRelease.yml]: should we rather use `win-latest`?

File: tools/releaseBuild/azureDevOps/vpackRelease.yml [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+  - template: ./templates/SetVersionVariables.yml
+    parameters:
+      ReleaseTagVar: $(ReleaseTagVar)
+
+  - powershell:  Write-Host "##vso[build.updatebuildnumber]$env:BUILD_SOURCEBRANCHNAME-$env:BUILD_SOURCEVERSION-$((get-date).ToString("yyyyMMddhhss"))"
+    displayName: Set Build Name for Non-PR
+    condition: ne(variables['Build.Reason'], 'PullRequest')
+
+- template: ./templates/vpackReleaseJob.yml
```
Comment (adityapatwardhan) [inline tools/releaseBuild/azureDevOps/vpackRelease.yml]: Should we consider:

1. Adding approvals?
2. Execute only on build.reason manual?

**Review by adityapatwardhan — APPROVED** [review]

File: tools/releaseBuild/azureDevOps/templates/vpackReleaseJob.yml (hunk 4 above)
Comment (TravisEz13) [inline tools/releaseBuild/azureDevOps/templates/vpackReleaseJob.yml]:
```suggestion
```

File: tools/releaseBuild/azureDevOps/vpackRelease.yml (hunk 5 above)
Comment (TravisEz13) [reply tools/releaseBuild/azureDevOps/vpackRelease.yml]:
```suggestion
    vmImage: 'windows-latest'
```

File: tools/releaseBuild/azureDevOps/templates/vpackReleaseJob.yml [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    inputs:
+      sourceDirectory: '$(vpackDir)'
+      description: PowerShell ${{ parameters.architecture }} $(version)
+      pushPkgName: 'PowerShell.${{ parameters.architecture }}'
+      configurations: Release
+      platforms: x64
+      target: '$(System.ArtifactsDirectory)'
+      owner: tplunk
+      provData: false
+      version: '$(vpackVersion)'
```
Comment (TravisEz13) [inline tools/releaseBuild/azureDevOps/templates/vpackReleaseJob.yml]:
```suggestion
      version: '$(vpackVersion)'
    condition: and(succeeded(), eq(variables['Build.Reason'], 'Manual'))
```

File: tools/releaseBuild/azureDevOps/vpackRelease.yml (hunk 6 above)
Comment (TravisEz13) [reply tools/releaseBuild/azureDevOps/vpackRelease.yml]: I did number 2

Let's leave #1 for when we integrate this into a deployment yaml.

**Pull request merged**